### PR TITLE
Register @ConfigMappings directly into the Config builder

### LINKED ISSUE #18333: Injecting ConfigMapping in resteasy containerrequestfilter causes NoSuchElementException
## Describe the bug
This one is quite similar to #17466, except this issue is about containerrequestfilter, while the other was about requestscoped observer.

```
@Provider
public class Filter implements ContainerRequestFilter {
    @Inject
    SomeConfigMapping config;

    @Override public void filter(ContainerRequestContext requestContext) throws IOException {
        System.out.println(config.value());
    }
}
```

```
@ConfigMapping(prefix = "some")
public interface SomeConfigMapping {
    @WithName("value")
    String value();
}
```

### Expected behavior
App should start without problems.

### Actual behavior
```
2021-07-02 08:32:01,404 INFO  [io.qua.dep.dev.IsolatedDevModeMain] (main) Attempting to start live reload endpoint to recover from previous Quarkus startup failure
2021-07-02 08:32:01,404 ERROR [io.qua.run.boo.StartupActionImpl] (Quarkus Main Thread) Error running Quarkus: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:98)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.lang.Class.newInstance(Class.java:584)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:65)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:42)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:119)
        at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
        ... 6 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:291)
        ... 15 more
Caused by: java.lang.RuntimeException: Error injecting org.acme.SomeConfigMapping org.acme.Filter.config
        at org.acme.Filter_Bean.create(Filter_Bean.zig:199)
        at org.acme.Filter_Bean.create(Filter_Bean.zig:222)
        at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
        at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
        at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
        at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
        at org.acme.Filter_Bean.get(Filter_Bean.zig:254)
        at org.acme.Filter_Bean.get(Filter_Bean.zig:270)
        at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:430)
        at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:443)
        at io.quarkus.arc.impl.ArcContainerImpl$1.get(ArcContainerImpl.java:266)
        at io.quarkus.arc.impl.ArcContainerImpl$1.get(ArcContainerImpl.java:263)
        at io.quarkus.resteasy.common.runtime.QuarkusConstructorInjector.construct(QuarkusConstructorInjector.java:39)
        at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.injectedInstance(ResteasyProviderFactoryImpl.java:1399)
        at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$AbstractInterceptorFactory.createInterceptor(JaxrsInterceptorRegistryImpl.java:150)
        at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$OnDemandInterceptorFactory.initialize(JaxrsInterceptorRegistryImpl.java:168)
        at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$OnDemandInterceptorFactory.checkInitialize(JaxrsInterceptorRegistryImpl.java:183)
        at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$OnDemandInterceptorFactory.getInterceptor(JaxrsInterceptorRegistryImpl.java:193)
        at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl$AbstractInterceptorFactory.postMatch(JaxrsInterceptorRegistryImpl.java:131)
        at org.jboss.resteasy.core.interception.jaxrs.JaxrsInterceptorRegistryImpl.postMatch(JaxrsInterceptorRegistryImpl.java:288)
        at org.jboss.resteasy.core.interception.jaxrs.ContainerRequestFilterRegistryImpl.postMatch(ContainerRequestFilterRegistryImpl.java:30)
        at org.jboss.resteasy.core.interception.jaxrs.ContainerRequestFilterRegistryImpl.postMatch(ContainerRequestFilterRegistryImpl.java:12)
        at org.jboss.resteasy.core.ResourceMethodInvoker.<init>(ResourceMethodInvoker.java:142)
        at org.jboss.resteasy.core.ResourceMethodRegistry.processMethod(ResourceMethodRegistry.java:381)
        at org.jboss.resteasy.core.ResourceMethodRegistry.register(ResourceMethodRegistry.java:308)
        at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:259)
        at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:227)
        at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:208)
        at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:192)
        at org.jboss.resteasy.core.ResourceMethodRegistry.addResourceFactory(ResourceMethodRegistry.java:175)
        at org.jboss.resteasy.core.ResourceMethodRegistry.addPerRequestResource(ResourceMethodRegistry.java:87)
        at org.jboss.resteasy.core.ResteasyDeploymentImpl.registerResources(ResteasyDeploymentImpl.java:518)
        at org.jboss.resteasy.core.ResteasyDeploymentImpl.registration(ResteasyDeploymentImpl.java:475)
        at org.jboss.resteasy.core.ResteasyDeploymentImpl.startInternal(ResteasyDeploymentImpl.java:164)
        at org.jboss.resteasy.core.ResteasyDeploymentImpl.start(ResteasyDeploymentImpl.java:121)
        at io.quarkus.resteasy.runtime.standalone.ResteasyStandaloneRecorder.staticInit(ResteasyStandaloneRecorder.java:37)
        at io.quarkus.deployment.steps.ResteasyStandaloneBuildStep$staticInit-210558872.deploy_0(ResteasyStandaloneBuildStep$staticInit-210558872.zig:903)
        at io.quarkus.deployment.steps.ResteasyStandaloneBuildStep$staticInit-210558872.deploy(ResteasyStandaloneBuildStep$staticInit-210558872.zig:40)
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:260)
        ... 15 more
Caused by: java.util.NoSuchElementException: SRCFG00027: Could not find a mapping for org.acme.SomeConfigMapping
        at io.smallrye.config.ConfigMappings.getConfigMapping(ConfigMappings.java:73)
        at io.smallrye.config.SmallRyeConfig.getConfigMapping(SmallRyeConfig.java:357)
        at io.quarkus.arc.runtime.ConfigMappingCreator.create(ConfigMappingCreator.java:28)
        at org.acme.SomeConfigMapping_2d5671bb9f49e16a6bde817d1dc2ee2e8da3440b_Synthetic_Bean.create(SomeConfigMapping_2d5671bb9f49e16a6bde817d1dc2ee2e8da3440b_Synthetic_Bean.zig:128)
        at org.acme.SomeConfigMapping_2d5671bb9f49e16a6bde817d1dc2ee2e8da3440b_Synthetic_Bean.get(SomeConfigMapping_2d5671bb9f49e16a6bde817d1dc2ee2e8da3440b_Synthetic_Bean.zig:159)
        at org.acme.SomeConfigMapping_2d5671bb9f49e16a6bde817d1dc2ee2e8da3440b_Synthetic_Bean.get(SomeConfigMapping_2d5671bb9f49e16a6bde817d1dc2ee2e8da3440b_Synthetic_Bean.zig:182)
        at io.quarkus.arc.impl.CurrentInjectionPointProvider.get(CurrentInjectionPointProvider.java:52)
        at org.acme.Filter_Bean.create(Filter_Bean.zig:182)
        ... 55 more

2021-07-02 08:32:01,467 INFO  [org.jbo.threads] (main) JBoss Threads version 3.4.0.Final
```

## To Reproduce

1. Download the reproducer: 
[requestscoped-test.zip](https://github.com/quarkusio/quarkus/files/6752574/requestscoped-test.zip)
2. mvn quarkus:dev
3. app does not start
4. Remove the filter class
5. mvn quarkus:dev
6. The app starts

### Configuration
```properties
some.value=abc
```

### Output of `uname -a` or `ver`
MSYS_NT-10.0 NANB7NLNVP2 2.10.0(0.325/5/3) 2018-06-13 23:34 x86_64 Msys

### Output of `java -version`
openjdk 11.0.7 2020-04-14 LTS
OpenJDK Runtime Environment Zulu11.39+15-CA (build 11.0.7+10-LTS)
OpenJDK 64-Bit Server VM Zulu11.39+15-CA (build 11.0.7+10-LTS, mixed mode)

### Quarkus version or git rev
2.0.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)
Maven home: C:\eclipse\tools\apache-maven\bin\..
Java version: 11.0.7, vendor: Azul Systems, Inc., runtime: C:\eclipse\tools\java\11
Default locale: de_DE, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

### PR DESCRIPTION
**Pull request by radcortez**
- Fixes #18333

Initial work to register the mappings directly with the builder.
- [x] Requires an update in SR Config (https://github.com/quarkusio/quarkus/pull/18629)
- [x] Requires the registration for static init (https://github.com/quarkusio/quarkus/pull/17483)

### CODE REVIEW

**Comment by gsmet** [comment]
I rebased it as it was a bit old.

**Review by gsmet — APPROVED** [review]

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/ConfigClassBuildItem.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package io.quarkus.arc.deployment;
+package io.quarkus.deployment.builditem;
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/ConfigClassBuildItem.java:1]: I wonder if we should just deprecate the old one for 2.2? Or we are sure it's not used anywhere in our extension ecosystem?

(deprecating a build item is easy, you just consume the old ones and produce the new ones)

Comment (radcortez) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/ConfigClassBuildItem.java:1]: I think it is ok to rename, this is only intended for internal usage.

This raises an interesting question. Do we have a way to distinguish which build step may be used safely on the ecosystem and which are for internal only? Unfortunately, a regular access modifier is not enough.

Comment (gsmet) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/ConfigClassBuildItem.java:1]: No, we don't.

**Comment by gsmet** [comment]
@radcortez can you please ping me here when this is ready, I keep forgetting about it. Thanks!

**Comment by radcortez** [comment]
> @radcortez can you please ping me here when this is ready, I keep forgetting about it. Thanks!

Yes, it is ready, I've just rebased with `main`.

**Comment by radcortez** [comment]
Yes

> @radcortez
> 
> ~Should this now also allow for constructor injection of configs into a an exception mapper? I was getting the same error as reported in #18333 with my exception mapper below.~
> 
> ```kotlin
> @Provider
> @ApplicationScoped
> class RouteExceptionInterceptor(config: MyConfig) : ExceptionMapper<Exception> {
> ```
> 
> ~Since I found this I bumped Quarkus to 2.2.0.CR1. The app still does not start but the config error is now different. Hence, I was wondering if this indicates this issue is not fully fixed yet OR whether I am having a different problem.~
> 
> > Should this now also allow for constructor injection of configs into a an exception mapper?
> 
> Yes, this PR fixes this as well. Thanks!

Yes, but unfortunately we need to revert this because it is creating other issues: https://github.com/quarkusio/quarkus/pull/19937

**Pull request merged**